neurons/validator.py

```diff
@@ -92,29 +97,50 @@ async def forward(self):
         )
         bt.logging.info(f"Received responses: {responses}")
 
-        rewards = self.validate_responses(responses)
+        rewards = self.validate_responses(responses, reference_report)
 
         bt.logging.info(f"Scored responses: {rewards}")
 
         self.update_scores(rewards, miner_uids)
 
-    def validate_responses(self, responses: List[AuditsSynapse]) -> List[float]:
-        return [self.validate_response(response) for response in responses]
-
-    def validate_response(self, response: AuditsSynapse) -> float:
-        if response:
-            validate_result = create_session().post(
-                f"{os.getenv('VALIDATOR_SERVER')}/validate",
-                json=jsonable_encoder(response.response),
-            )
-
-            if validate_result.status_code != 200:
-                bt.logging.error("Miner synapse is incorrect. Scored reward is 0")
-                return 0.0
-            else:
-                return validate_result.json().get("result", 0.0)
+    def validate_responses(
+        self,
+        responses: List[AuditsSynapse],
+        reference_report: List[VulnerabilityReport] = [],
+    ) -> List[float]:
+        return [
+            self.validate_reports_by_reference(response.response, reference_report)
+            for response in responses
+        ]
+
+    def validate_reports_by_reference(
+        self,
+        report: List[VulnerabilityReport] | None,
+        reference_report: List[VulnerabilityReport],
+    ) -> float:
+        if report is None:
+            return 0.0
+
+        finded_vulns = [vuln.vulnerability_class for vuln in report]
+        reference_vulns = [vuln.vulnerability_class for vuln in reference_report]
+        finded_count = Counter(finded_vulns)
+        reference_count = Counter(reference_vulns)
+
+        intersection = [
+            vuln
+            for vuln in finded_vulns
+            for _ in range(min(finded_count[vuln], reference_count[vuln]))
+        ]
+
+        return len(intersection) / len(reference_vulns)
```

Comment:
The code can be simplified a bit here, as Python allows for that.

```suggestion
    def validate_reports_by_reference(
        self,
        report: Optional[List[VulnerabilityReport]],
        reference_report: List[VulnerabilityReport],
    ) -> float:
        if report is None or not reference_report:
            return 0.0

        finded_count = Counter(vuln.vulnerability_class for vuln in report)
        reference_count = Counter(vuln.vulnerability_class for vuln in reference_report)

        intersection_count = finded_count & reference_count

        return intersection_count / len(reference_report)
```